### DISCUSSION TITLE
What kind of API do we need?

### DISCUSSION THREAD

**Discussion by kyrreso** [Norsk-institutt-for-samkjoring/Standard#2] [Polls]
Can you help us? Have a look at https://github.com/Norsk-institutt-for-samkjoring/Standard/blob/main/OpenCarPoolingStandard.md

**Comment by kyrreso** [comment]
we ale have the one from https://github.com/mitfahrverband/mitfahr-api/tree/main

**Comment by kyrreso** [comment]
and we have  https://github.com/openmobilityfoundation/mobility-data-specification

https://github.com/mfdz/amarillo

**Comment by MartinH-open** [comment]
We need to define the domain (the context, scope) for which we need data / api standards.
Is it about what actual ride sharing offers are available (when from where to where) and the life cycle of such an offer (create, read, update, delete)?
Is it about managing the booking, executing, confirming and reproting ride sharing trips?
Is it about requesting a ride share (searching for a ride share) and notifications of suitable ride sharing matches?
Is it about ride sharing user management, user ratings or managing related payment transactions?

**Comment by kyrreso** [comment]
Hmm. I like how you separate it. If possible, I would hope for us to cover all things, and later define what is needed for different apps/solution to talk to each other, and do the whole routine with another from another eco-system. It could also just be a whitepaper for new apps, but the best for customers would be that the same thing should work if one users changes onto another app, and do a ride the day after. This would only work if the apps follow the standard. It could also be a requirement for the one competing for funds to follow any open standard like this, at least use API from theirs.

**Comment by MartinH-open** [comment]
Model: 
The minimum set of real world objects that are required for a carpooling are a "car" (vehicle) and a "pool" of at least two people. Most of the time at least one person brings one car along (but well, you can imaging two people meeting somewhere to pick up a carsharing vehicle to build their carpool). Each carpooling instance has some life cycle: it starts with some demand for mobility at one person (having a car or not). The demand has some point in space and time to start from and most times some (different) destination point in space and time. Multiple of such demands match to each other more or less. To evaluate the matches the current demand data has to be "joined" somehow (come together in one database for match evaluation.) There might be a number of factors, additional attributes the determine how good the match of two demands evaluates. Today most carpooling services manage a combination of a person and a car as a "driver" who is offering rides from A to B. And on the other hand are persons (without car) who search for rides for C to D.  Both sides have certain time constraints for their rides (from just now to any time in future). 
 .. and so on.

Are we looking for the most generic model for carpooling ?
Or do we look for exchange / integration of multiple carpooling services to find a common language and API to exchange carpooling related data?
Is the purpose to have some reporting only? Do we want to optimize matching between many / all integrated parties?
For designing a good API we need to understand the purposes and stories we want to support. 
Perhaps we need some event storming session .. ?

**Reply by kyrreso** [reply]
The hope and plan for the standard:

1.    Create the most advanced and comprehensive dataset for new apps to utilize, inspiring these companies and facilitating the generation of innovative ideas for inclusion. It would serve as a comprehensive guide outlining what they can or should incorporate.
3.    Establish a foundation for a standard that can serve as a source of inspiration for other standards, while also establishing requirements for apps to facilitate better communication with each other.
4.    Demonstrate to municipalities and Public Transport Authorities (PTAs) that they can establish requirements and initiate negotiations with carpooling app suppliers to adopt a standard and collaborate with multiple providers.

**Reply by MartinH-open** [reply]
I understand the needs for (2) and (3) to some extend.
An open standard for car pooling should help to integrate services with PTA and to enable cooperation between carpooling agencies.
(0) So one subject is to describe (publish) information about carpooling agencies their available services.
(A) Other subjects are the services for accepting, updating/removing, publishing individual carpooling rides (offers and demands). 
(B) Next could be request/reserve, book, accept a ride and to negotiate and accompany to a meeting point, starting or canceling a ride. 
(C) Escorting a ride with all possibile events during the ride from having fun to escaping from the vehicle; together with documenting the common journey, 
(D) signaling the carpooling ride to the infrastructure (like HOV lanes, parking places, city zones, ..). 
(E) Other topics might be payments, benefits, ticketing or compensation.
(F) Finally giving ratings to eachother and connect or disconnect. 
(G) Finally if needed allow investigation of incidents.
(H) In total there is some need for general reporting to the public (as open data) - to measure success and justify any investments.
(X) For all user related topics we might want some user management services to provide identity managment across services.

As of now I see the main focus has been on (A) and some basic information management of type (0) and (H) data.

**Reply by kyrreso** [reply]
Thank you a lot for the feedback, I have now implemented many of your points in the 0.9, Have a look! -Kyrre
